### ISSUE TITLE
Invalid data field

### ISSUE THREAD

**Issue by 1smii** [ferozed/mask-json-field-transform#1]
Hello, 

I have downloaded the jar files to use them in my project.

My use case looks like this:
I have a CDC architecture to capture all data changes from a PostgreSQL Database, meanwhile using Debezium Connector, I have installed the following plugin https://github.com/RedHatInsights/expandjsonsmt to expand a field that contains a JSON object inside a string.

After the transformation to expand the JSON object, I'm using your mask-json-field-transform following the documentation provided. Check a piece of my connector config:

      "transforms": "unwrap,expand,mask",
      "transforms.unwrap.type": "io.debezium.transforms.ExtractNewRecordState",
      "transforms.unwrap.drop.tombstones": "false",
      "transforms.expand.type": "com.redhat.insights.expandjsonsmt.ExpandJSON$Value",
      "transforms.expand.sourceFields": "basket",
      "transforms.mask.type": "io.github.ferozed.kafka.connect.transforms.MaskJsonField$Value",
      "transforms.mask.REPLACEMENT_FIELD_PATH": "/basket/locale",
      "transforms.mask.REPLACEMENT_VALUE_STRING": "******",

The expanded basket field contains multiples child, one of them is /basket/locale.
I can guarantee the fields exits, however, I'm getting some errors:
ERROR  ||  WorkerSourceTask{id=basket-connector-1-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted   [org.apache.kafka.connect.runtime.WorkerTask]
debezium-connect  | org.apache.kafka.connect.errors.ConnectException: Tolerance exceeded in error handler
debezium-connect  | 	at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:230)
debezium-connect  | 	at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execute(RetryWithToleranceOperator.java:156)
debezium-connect  | 	at org.apache.kafka.connect.runtime.TransformationChain.apply(TransformationChain.java:50)
debezium-connect  | 	at org.apache.kafka.connect.runtime.AbstractWorkerSourceTask.sendRecords(AbstractWorkerSourceTask.java:391)
debezium-connect  | 	at org.apache.kafka.connect.runtime.AbstractWorkerSourceTask.execute(AbstractWorkerSourceTask.java:359)
debezium-connect  | 	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:202)
debezium-connect  | 	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:257)
debezium-connect  | 	at org.apache.kafka.connect.runtime.AbstractWorkerSourceTask.run(AbstractWorkerSourceTask.java:75)
debezium-connect  | 	at org.apache.kafka.connect.runtime.isolation.Plugins.lambda$withClassLoader$1(Plugins.java:177)
debezium-connect  | 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
debezium-connect  | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
debezium-connect  | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
debezium-connect  | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
debezium-connect  | 	at java.base/java.lang.Thread.run(Thread.java:829)
debezium-connect  | Caused by: org.apache.kafka.connect.errors.DataException:  is not a valid field name
debezium-connect  | 	at org.apache.kafka.connect.data.Struct.lookupField(Struct.java:254)
debezium-connect  | 	at org.apache.kafka.connect.data.Struct.getCheckType(Struct.java:261)
debezium-connect  | 	at org.apache.kafka.connect.data.Struct.getString(Struct.java:158)
debezium-connect  | 	at io.github.ferozed.kafka.connect.transforms.MaskJsonField.processStruct(MaskJsonField.java:154)
debezium-connect  | 	at myapp.com.github.jcustenborder.kafka.connect.transform.common.BaseTransformation.process(BaseTransformation.java:131)
debezium-connect  | 	at io.github.ferozed.kafka.connect.transforms.MaskJsonField.apply(MaskJsonField.java:78)
debezium-connect  | 	at org.apache.kafka.connect.runtime.TransformationChain.lambda$apply$0(TransformationChain.java:50)
debezium-connect  | 	at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndRetry(RetryWithToleranceOperator.java:180)
debezium-connect  | 	at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:214)
debezium-connect  | 	... 13 more
debezium-connect  | 2023-08-03 09:41:44,008 INFO   ||  Stopping down connector   [io.debezium.connector.common.BaseSourceTask]

I have tried to use different ways as you commented on the README.MD without success.

It is possible to use your plugin in this case?

I hope I went clear with this!

Regards
Ismael

**Comment by ferozed** [comment]
Can you give me the full connector config? You can mask out passwords and other sensitive data.

Also, give me an example payload that failed. Again, mask out the sensitive data.

**Comment by 1smii** [comment]
Hello ferozed,

Sure!
This is my source connector:

```
{
  "name": "basket-connector-1",
  "config": {
      "connector.class": "io.debezium.connector.postgresql.PostgresConnector",
      "tasks.max": "1",
      "database.hostname": "test",
      "database.port": "5432",
      "database.user": "test",
      "database.password": "test",
      "database.dbname" : "shop",
      "topic.prefix": "example_topic",
      "table.include.list": "schema_name.basket_shop",
      "key.converter":  "org.apache.kafka.connect.json.JsonConverter",
      "value.converter": "org.apache.kafka.connect.json.JsonConverter",
      "value.converter.schemas.enable": "false",
      "key.converter.schemas.enable": "false",
      "transforms": "unwrap,expand",
      "transforms.unwrap.type": "io.debezium.transforms.ExtractNewRecordState",
      "transforms.unwrap.drop.tombstones": "false",
      "transforms.expand.type": "com.redhat.insights.expandjsonsmt.ExpandJSON$Value",
      "transforms.expand.sourceFields": "basket",
      "transforms.mask.type": "io.github.ferozed.kafka.connect.transforms.MaskJsonField$Value",
      "transforms.mask.REPLACEMENT_FIELD_PATH": "/basket/locale",
      "transforms.mask.REPLACEMENT_VALUE_STRING": "******",
      "slot.name": "basketconnector1"
  }
}
```

There is a column in the database called "basket" which is a JSON Object inside a string. That's why I'm using expandjsonsmt to be able to access all the childs of the basket field.

In the end, the idea is to mask some fields of the expanded basket field.
This is an example of the input:

```
{
  "basket_id": "123918209",
  "order_no": "a29jas",
  "basket": {
    "basket_id": "123918209",
    "locale": "EU",
    "status": {
      "orderStatus": "NEW",
      "paymentStatus": "NEW"
    },
    "basket_items": [
      {
        "basket_item_id": "123918209",
        "product_id": "123918209",
        "quantity": 1,
        "price": 100,
        "currency": "USD"
      }
    ],
    "total_price": 100,
    "currency": "USD"
  }
}
```

I'm not able to mask any field, for example, these:
order_no
basket.locale
basket.basket_items.0.product_id

I had some errors, I am not able to provide the error logs today, I'll do it tomorrow.

Some errors like:
Use "/": "STRING" (when I was using it)
Not able to find the field "basket"
The field: (empty) is not assignable

Etc.

I hope you can try these examples! I'll keep you posted with the error messages tomorrow!

Kindly Regards,
Ismael

**Comment by ferozed** [comment]
Hmm... Not understanding your use case.

You are using `JsonConverter`. That means that the data in your topic is JSON? With JSONConverter, the JSON message will be converted into connect internal schema that will closely match the incoming message.

If that is the case, should'nt you just use other transforms to mask the field?

The `mask-json-field` transform works on JSON payloads that come as strings. In other words, if you didnt use a JsonConverter, and used a StringConverter instead, then you could use this transform.

But with JSONConverter, and the way the message is coming on the topic, you cant.

**Comment by ferozed** [comment]
Also, in the connect config you gave me, the mask-json transform is just not being used.

`     
"transforms": "unwrap,expand",
`

**Comment by ferozed** [comment]
The other option is to change the order of the transforms. I am not sure what unwrap is doing, but if you could run the mask json field transform first, then unwrap and expand, it might work,

**Comment by 1smii** [comment]
Hello ferozed,

Sorry for not explaining myself correctly,
let me elaborate a little better:

I have a database, with different columns and one of them (basket) is of type jsonb which means its value is a string with a json object, others are not json.

The idea of using jsonConverter is to have all columns with JSON key/value.

The transform "unwrap" io.debezium.transforms.ExtractNewRecordState is used to only capture the changes that occur in the table, updates, inserts, deletes, the "transforms.unwrap.drop.tombstones": "false", is to control the "delete" records and that these are still sent to the final topic.
When using 
```
      "transforms.expand.type": "com.redhat.insights.expandjsonsmt.ExpandJSON$Value",
      "transforms.expand.sourceFields": "basket"
```

to be able to access all the fields that were inside the string.

PS: I forgot to put
```
      "transforms": "unwrap,expand,mask"
```

But if I had it

**Comment by 1smii** [comment]
All this connector is to expand that jsonb field (basket), the mask must be applied to one of the fields in the basket column, it doesn't matter if it's before using expand or after.

My problem is that I don't know how to apply this mask to this field. Since I use jsonConverter as you said I should use stringConverter, use the mask and finally somehow have it all in key/value json.

**Comment by 1smii** [comment]
Also, I have tried to use maskField from confluent without success https://docs.confluent.io/platform/current/connect/transforms/maskfield.html

It can't enter the expanded JSON structure

**Comment by ferozed** [comment]
Hmm... i still dont understand.

Can you read one message from the input topic, and show me? Mask out sensitive values.

**Comment by ferozed** [comment]
Ok... while you are waiting to give me a raw message from the topic, I looked at the JSONConverter.

So... what is happening, is that since you specified "schema.enable=false" the JSONconverter is creating an internal object of type `MAP` instead of `Struct`. And my converter only supports `Structs` and `Strings` right now.

Can you tell me what is the goal of your converter? Are you trying to strip PII data and send the payload into a topic as JSON?

**Comment by 1smii** [comment]
Hello Ferozed,

The goal of this converter si to have the payload into a topic as JSON with some fields masked.

**Comment by ferozed** [comment]
``I see. In that case, you dont need to run so many transformers with JSONConverter.

1. Instead of JSONConverter, use StringConverter for value
2. Only use the `maskJsonFieldTransform` to mask the basket fields

Example:

```
"value.converter": "org.apache.kafka.connect.storage.StringConverter",
"transforms": "mask_json_field",
"transforms.mask_json_field.type": "io.github.ferozed.kafka.connect.transforms.MaskJsonField$Value",
"transforms.mask_json_field.REPLACEMENT_FIELD_PATH": "/basket/locale",

```

Give that a try and see if it works. 

If you need to mask more fields, just add more transforms for the extra fields.

Meanwhile I will try to implement support for MAP types as well.

**Comment by 1smii** [comment]
Hi Ferozed,

I have tried your configuration without success.

This is one of the errors while creating the connector:
```
debezium-connect  | 2023-08-08 06:47:49,321 ERROR  ||  WorkerSourceTask{id=basket-connector-1-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted   [org.apache.kafka.connect.runtime.WorkerTask]
debezium-connect  | org.apache.kafka.connect.errors.ConnectException: Tolerance exceeded in error handler
debezium-connect  | 	at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:230)
debezium-connect  | 	at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execute(RetryWithToleranceOperator.java:156)
debezium-connect  | 	at org.apache.kafka.connect.runtime.TransformationChain.apply(TransformationChain.java:50)
debezium-connect  | 	at org.apache.kafka.connect.runtime.AbstractWorkerSourceTask.sendRecords(AbstractWorkerSourceTask.java:391)
debezium-connect  | 	at org.apache.kafka.connect.runtime.AbstractWorkerSourceTask.execute(AbstractWorkerSourceTask.java:359)
debezium-connect  | 	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:202)
debezium-connect  | 	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:257)
debezium-connect  | 	at org.apache.kafka.connect.runtime.AbstractWorkerSourceTask.run(AbstractWorkerSourceTask.java:75)
debezium-connect  | 	at org.apache.kafka.connect.runtime.isolation.Plugins.lambda$withClassLoader$1(Plugins.java:177)
debezium-connect  | 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
debezium-connect  | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
debezium-connect  | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
debezium-connect  | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
debezium-connect  | 	at java.base/java.lang.Thread.run(Thread.java:829)
debezium-connect  | Caused by: org.apache.kafka.connect.errors.DataException:  is not a valid field name
debezium-connect  | 	at org.apache.kafka.connect.data.Struct.lookupField(Struct.java:254)
debezium-connect  | 	at org.apache.kafka.connect.data.Struct.getCheckType(Struct.java:261)
debezium-connect  | 	at org.apache.kafka.connect.data.Struct.getString(Struct.java:158)
debezium-connect  | 	at io.github.ferozed.kafka.connect.transforms.MaskJsonField.processStruct(MaskJsonField.java:154)
debezium-connect  | 	at myapp.com.github.jcustenborder.kafka.connect.transform.common.BaseTransformation.process(BaseTransformation.java:131)
debezium-connect  | 	at io.github.ferozed.kafka.connect.transforms.MaskJsonField.apply(MaskJsonField.java:78)
debezium-connect  | 	at org.apache.kafka.connect.runtime.TransformationChain.lambda$apply$0(TransformationChain.java:50)
debezium-connect  | 	at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndRetry(RetryWithToleranceOperator.java:180)
debezium-connect  | 	at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:214)
debezium-connect  | 	... 13 more
```

This is my connector config:

```
{
  "name": "basket-connector-1",
  "config": {
      "connector.class": "io.debezium.connector.postgresql.PostgresConnector",
      "tasks.max": "1",
      "database.hostname": "test",
      "database.port": "5432",
      "database.user": "test",
      "database.password": "test",
      "database.dbname" : "baskets",
      "topic.prefix": "test-topic",
      "table.include.list": "public.basket_entity",
      "key.converter": "org.apache.kafka.connect.storage.StringConverter",
      "value.converter": "org.apache.kafka.connect.storage.StringConverter",
      "value.converter.schemas.enable": "false",
      "key.converter.schemas.enable": "false",
      "transforms": "mask_json_field",
      "transforms.mask_json_field.type": "io.github.ferozed.kafka.connect.transforms.MaskJsonField$Value",
      "transforms.mask_json_field.REPLACEMENT_FIELD_PATH": "/basket/locale",
      "slot.name": "basketconnector1"
  }
}
```

These are my db column types
![image](https://github.com/ferozed/mask-json-field-transform/assets/48752102/d0e9933f-cf60-43d1-aec6-6ab56998b343)

Another thing to have in mind is while using expandjsonsmt plugin, this plugin expands a key whose value is a string with a JSON Object.

That is the reason to have all fields key/value as JSON type.

**Comment by 1smii** [comment]
The idea is to normalize the JSONB "basket" field into a JSON to be able to access all their fields via a data browser or ElasticSearch Index.

The final goal looks like this:
![image](https://github.com/ferozed/mask-json-field-transform/assets/48752102/ee8721ee-9441-4dd7-9032-82c0befc856c)

**Comment by ferozed** [comment]
Is `jsonb` binary json? Is there any way you can convert it into a JSON string? Then you can use my transform.

**Comment by ferozed** [comment]
hi! @1smii . Can you try this fix?

https://github.com/ferozed/mask-json-field-transform/pull/2

**Comment by 1smii** [comment]
Hi @ferozed ,

Excuse me for the delay, related to the jsonb field, it's binary json, you're right. I Can convert it into a JSON string, and try to use the transform, There is just one doubt, do you support arrays? 

The main issue is that in the JSON String field are different arrays, which I'll need to expand later.

**Comment by 1smii** [comment]
I'll try this fix!

> hi! @1smii . Can you try this fix?
> 
> #2

**Comment by ferozed** [comment]
Hi!

Yeah, if you can convert it into a string then you can use the old version as well. The key thing is that it has to be a json string. Otherwise there are other transformers that would work as sep

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by 1smii** [comment]
Hi @ferozed,

I have been doing some tests to convert it into a string and it's working as expected!

I'm going to test these structure control features, however, I'll create another ticket in that case.

From my end, this issue is solved.

Kindly Regards,
Ismael

**Comment by ferozed** [comment]
awesome! can you show me the connector config that worked?
________________________________

**Comment by ferozed** [comment]
I released `1.1` with support for maps as well.
